crates/bevy_math/src/direction.rs

```diff
@@ -866,6 +867,251 @@ impl approx::UlpsEq for Dir3A {
+    ///
+    /// Returns [`Err(InvalidDirectionError)`](InvalidDirectionError) if the length
+    /// of the given vector is zero (or very close to zero), infinite, or `NaN`.
+    pub fn new_and_length(value: Vec4) -> Result<(Self, f32), InvalidDirectionError> {
+        let length = value.length();
+        let direction = (length.is_finite() && length > 0.0).then_some(value / length);
+
+        direction
+            .map(|dir| (Self(dir), length))
+            .ok_or(InvalidDirectionError::from_length(length))
+    }
+
+    /// Create a direction from its `x`, `y`, `z`, and `w` components.
+    ///
+    /// Returns [`Err(InvalidDirectionError)`](InvalidDirectionError) if the length
+    /// of the vector formed by the components is zero (or very close to zero), infinite, or `NaN`.
+    pub fn from_xyzw(x: f32, y: f32, z: f32, w: f32) -> Result<Self, InvalidDirectionError> {
+        Self::new(Vec4::new(x, y, z, w))
+    }
+
+    /// Create a direction from its `x`, `y`, `z`, and `w` components, assuming the resulting vector is normalized.
+    ///
+    /// # Warning
+    ///
+    /// The vector produced from `x`, `y`, `z`, and `w` must be normalized, i.e its length must be `1.0`.
+    pub fn from_xyzw_unchecked(x: f32, y: f32, z: f32, w: f32) -> Self {
+        Self::new_unchecked(Vec4::new(x, y, z, w))
+    }
+
+    /// Returns the inner [`Vec4`]
+    pub const fn as_vec4(&self) -> Vec4 {
+        self.0
+    }
+
+    /// Performs a spherical linear interpolation between `self` and `rhs`
+    /// based on the value `s`.
+    ///
+    /// This corresponds to interpolating between the two directions at a constant angular velocity.
+    ///
+    /// When `s == 0.0`, the result will be equal to `self`.
+    /// When `s == 1.0`, the result will be equal to `rhs`.
+    ///
+    /// If the angle of between `self` and `rhs` is invalid it returns `self`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_math::Dir4;
+    /// # use approx::{assert_relative_eq, RelativeEq};
+    /// #
+    /// let dir1 = Dir4::W;
+    /// let dir2 = Dir4::Y;
+    ///
+    /// let result1 = dir1.slerp(dir2, 1.0 / 3.0);
+    /// #[cfg(feature = "approx")]
+    /// assert_relative_eq!(
+    ///     result1,
+    ///     Dir4::from_xyzw(0.0, 0.5, 0.0, 0.75_f32.sqrt()).unwrap(),
+    ///     epsilon = 0.000001
+    /// );
+    ///
+    /// let result2 = dir1.slerp(dir2, 0.5);
+    /// #[cfg(feature = "approx")]
+    /// assert_relative_eq!(result2, Dir4::from_xyzw(0.0, 0.5_f32.sqrt(), 0.0, 0.5_f32.sqrt()).unwrap());
+    /// ```
+    #[inline]
+    pub fn slerp(self, rhs: Self, s: f32) -> Self {
+        // This uses a geometric slerp as opposed to the Quaternion slerp Dir3 uses.
+        // When the values are too small to use slerp it falls back on a regular lerp.
+        let angle = acos(self.dot(*rhs));
+        if angle.is_nan() {
+            return self;
+        }
+        if angle == 0.0 || angle == PI {
```

Comment:
Exact float comparison doesn't help much here. While it avoids the case sin(0.0) == 0.0, it doesn't avoid all of the nearby values where the value is very small and the result of the division is a fairly absurdly large number.